contracts/DomainRegistry.sol

```diff
@@ -11,55 +16,99 @@ contract DomainRegistry {
+        require(domains[domainName].owner == msg.sender, "Not the domain owner");
         _;
     }
 
-    modifier domainOwnedBySender(string memory _topLevelDomain) {
-        require(domains[_topLevelDomain].owner == msg.sender, "Not the domain owner");
+modifier parentDomainExists(string memory domainName) {
+    if (isTopLevelDomain(domainName)) {
         _;
+        return;
+    }
+    
+    string memory parentDomain = domainName.extractParentDomain();
+    if (bytes(parentDomain).length > 1) {  
+        require(domains[parentDomain].isRegistered, "Parent domain doesn't exist");
     }
+    _;
+}
 
-    modifier isTopLevelDomain(string memory _topLevelDomain) {
-        require(bytes(_topLevelDomain).length > 0, "Domain is empty");
-        for (uint i = 0; i < bytes(_topLevelDomain).length; i++) {
-            require(bytes(_topLevelDomain)[i] != bytes(".")[0], "Multilevel domains are not allowed");
+function isTopLevelDomain(string memory domainName) internal pure returns (bool) {
+    bytes memory domainBytes = bytes(domainName);
+    uint dotCount = 0;
+    for (uint i = 0; i < domainBytes.length; i++) {
+        if (domainBytes[i] == bytes1('.')) {
+            dotCount++;
         }
-        _;
     }
+    return dotCount == 0;  
+}
 
-    function registerDomain(string memory _topLevelDomain) public hasRequiredDeposit() domainDoesNotExist(_topLevelDomain) isTopLevelDomain(_topLevelDomain) payable {
-        domains[_topLevelDomain] = Domain({
-            deposit: msg.value,
-            isRegistered: true,
-            owner: msg.sender
-        });
+function registerDomain(string memory domainName) 
+    public 
+    hasRequiredDeposit() 
+    domainExistence(domainName, false) 
+    isValidDomain(domainName)
+    parentDomainExists(domainName)
+    payable 
+{
+    domainName = domainName.stripProtocol();
+    require(bytes(domainName).length > 0, "Domain cannot be empty");
 
-        emit DomainCreated(_topLevelDomain, msg.value, msg.sender);
-    }
+    domains[domainName] = Domain({
+        deposit: msg.value,
+        isRegistered: true,
+        owner: msg.sender
+    });
+
+    emit DomainCreated(domainName, msg.value, msg.sender);
+}
 
-    function releaseDomain(string memory _topLevelDomain) public domainExists(_topLevelDomain) domainOwnedBySender(_topLevelDomain) {
-        Domain storage domain = domains[_topLevelDomain];
+    function releaseDomain(string memory domainName) 
+        public 
+        domainExistence(domainName, true) 
+        domainOwnedBySender(domainName) 
+    {
+        domainName = domainName.stripProtocol();
+        
+        Domain storage domain = domains[domainName];
         uint256 depositAmount = domain.deposit;
+
         domain.isRegistered = false;
         domain.deposit = 0;
         domain.owner = address(0);
+
         payable(msg.sender).transfer(depositAmount);
-        emit DomainReleased(_topLevelDomain);
+
+        emit DomainReleased(domainName);
```

Comment:
It's possible to release the domain without checking if it has some child domains. -1 point